### ISSUE TITLE
Performance when repeating circuit execution

### ISSUE THREAD

**Issue by stavros11** [Quantum-TII/qibo#272]
There are some inconsistencies in simulation time when executing circuits repeatedly on GPU. More specifically I tried the following benchmark:

```Python
thetas = np.random.random(nqubits)
circuit = models.Circuit(nqubits)
circuit.add((gates.RX(i, theta=t) for i, t in enumerate(thetas)))

start_time = time.time()
for _ in range(nreps):
    state = circuit()
    del(state)
total_time = time.time() - start_time
```
and I get the following numbers on CPU:

**10 qubits - complex128 - CPU 40 threads**
nreps | total time | time / nreps
-- | -- | --
1 | 0.00219 | 0.00219
25 | 0.03942 | 0.00158
50 | 0.07245 | 0.00145
75 | 0.11122 | 0.00148
100 | 0.13446 | 0.00134
250 | 0.36209 | 0.00145
500 | 0.73414 | 0.00147
750 | 1.19409 | 0.00159
1000 | 1.54387 | 0.00154
1500 | 2.31893 | 0.00155
2000 | 3.20498 | 0.00160

**20 qubits - complex128 - CPU 40 threads**
nreps | total time (sec) | time / nreps
-- | -- | --
1 | 0.01470 | 0.01470
25 | 0.23724 | 0.00949
50 | 0.44128 | 0.00883
75 | 0.65118 | 0.00868
100 | 0.85110 | 0.00851
250 | 2.11610 | 0.00846
500 | 4.10390 | 0.00821
750 | 6.07117 | 0.00809
1000 | 8.02729 | 0.00803

**25 qubits - complex128 - CPU 40 threads**
nreps | total time | time / nreps
-- | -- | --
1 | 0.55971 | 0.55971
5 | 2.79272 | 0.55854
10 | 5.56048 | 0.55605
15 | 8.38124 | 0.55875
20 | 11.16901 | 0.55845
25 | 13.94736 | 0.55789
30 | 16.82672 | 0.56089

Notice that in all cases the time / nreps ratio is constant, that is execution time increases linearly with nreps, as expected since we are executing the circuit sequentially. However, when executing on GPU I get the following:

**20 qubits - complex128 - V100 32GB**
nreps | total time | time / nreps
-- | -- | --
1 | 0.00546 | 0.00546
25 | 0.02110 | 0.00084
50 | 0.03800 | 0.00076
75 | 0.04793 | 0.00064
100 | 0.06691 | 0.00067
250 | 0.24763 | 0.00099
500 | 0.53438 | 0.00107

**25 qubits - complex128 - V100 32GB**
nreps | total time | time / nreps
-- | -- | --
1 | 0.04529 | 0.04529
25 | 0.06846 | 0.00274
50 | 0.53088 | 0.01062
75 | 1.61458 | 0.02153
100 | 2.69932 | 0.02699
250 | 9.18972 | 0.03676
500 | 19.98812 | 0.03998

**27 qubits - complex128 - V100 32GB**
nreps | total time | time / nreps
-- | -- | --
1 | 0.18310 | 0.18310
25 | 0.21440 | 0.00858
50 | 2.75848 | 0.05517
75 | 7.42070 | 0.09894
100 | 12.06227 | 0.12062
250 | 39.94245 | 0.15977
500 | 86.39704 | 0.17279

**28 qubits - complex128 - V100 32GB**
nreps | total time | time / nreps
-- | -- | --
1 | 0.37615 | 0.37615
25 | 0.39985 | 0.01599
50 | 6.16452 | 0.12329
75 | 15.80753 | 0.21077
100 | 25.43888 | 0.25439
250 | 83.23298 | 0.33293
500 | 179.56643 | 0.35913
1000 | 372.25714 | 0.37226
2000 | 757.57767 | 0.37879

**29 qubits - complex128 - V100 32GB**
nreps | total time | time / nreps
-- | -- | --
1 | 0.77786 | 0.77786
25 | 0.80320 | 0.03213
50 | 13.60502 | 0.27210
75 | 33.50419 | 0.44672
100 | 53.41602 | 0.53416
250 | 172.88784 | 0.69155
500 | 371.98466 | 0.74397
1000 | 770.17064 | 0.77017
2000 | 1566.86891 | 0.78343

so the single-rep time increases when we increase nreps beyond a specific number (around 50). Moreover, while it is possible to simulate a single rep of 30 qubits on the 32GB GPU, it runs OOM when `nreps > 1`.

All these benchmarks are using the custom operators. I still need to check if the behavior using only Tensorflow primitives.

**Comment by stavros11** [comment]
Here are some additional benchmarks using the `defaulteinsum` backend:

**27 qubits - complex128 - V100 32GB**
nreps | total time | time / nreps
-- | -- | --
1 | 1.12613 | 1.12613
25 | 16.88318 | 0.67533
50 | 33.31521 | 0.66630
75 | 49.74536 | 0.66327
100 | 66.18449 | 0.66184
250 | 164.75608 | 0.65902
500 | 329.44812 | 0.65890

**28 qubits - complex128 - V100 32GB**
nreps | total time | time / nreps
-- | -- | --
1 | 1.88750 | 1.88750
25 | 34.09706 | 1.36388
50 | 67.75381 | 1.35508
75 | 101.40478 | 1.35206
100 | 135.04928 | 1.35049
250 | 337.25821 | 1.34903
500 | 674.09125 | 1.34818

Here the situation is similar to CPU and as expected the nreps / time ratio is constant. Therefore the issue appears only when using the GPU with custom operators.

**Comment by scarrazza** [comment]
Yes, I agree, there is something odd for the GPU, e.g. I tried your code on the titan v, and there is something I don't understand, nreps=1, nqubits=28, I get:
 - 0.41259002685546875 for a single run
 - 0.0013480186462402344, when repeating the loop manually (copy paste twice the for loop)
the wave-function is correct, so there must be some caching algorithm.

**Comment by scarrazza** [comment]
We did some checks and we conclude the following:
- the OOM on GPU derives from the stored `_final_state` object inside the `Circuit` class. This object should be automatically deleted when executing the circuit.
- the sample code above does not contain an endpoint, thus tf caches and copies the objects, which may raise inconsistent repetition execution times.

**Comment by scarrazza** [comment]
Fixed in #274.
